### PR TITLE
[SPARK-48413][SQL][FOLLOW-UP] Fix ALTER COLUMN with collation

### PR DESCRIPTION
**Pull request by jovanm-db**
### What changes were proposed in this pull request?
Fixing the analysis check error introduced in [this](https://github.com/apache/spark/pull/46734) PR, where the `ALTER COLUMN` command would fail on tables that contain catalog in their names, ensuring consistent behavior across all table naming conventions. 

### Why are the changes needed?
Change is needed to enable altering column with all datasources and identifier variations. 

### Does this PR introduce _any_ user-facing change?
No. 

### How was this patch tested?
Added tests to `CollationSuite`. 

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 1]
```diff
@@ -1546,7 +1546,10 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog with QueryErrorsB
             .map(dt => col.field.copy(dataType = dt))
             .getOrElse(col.field)
           val newDataType = a.dataType.get
+          val sameTypeExceptCollations =
+            DataType.equalsIgnoreCompatibleCollation(field.dataType, newDataType)
           newDataType match {
+            case _ if sameTypeExceptCollations =>
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:1552]: Add comment saying as to why this is allowed. We allow it as we want to be able to change collations.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 2]
```diff
@@ -516,6 +516,14 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
     }
   }
 
+  test("SPARK-48413: Alter column with collation") {
+    val tableName = "testcat.alter_column_tbl"
+    withTable(tableName) {
+      sql(s"CREATE TABLE $tableName (c1 STRING) USING PARQUET")
+      sql(s"ALTER TABLE $tableName ALTER COLUMN c1 TYPE STRING COLLATE UTF8_LCASE")
+    }
```
Comment (mihailom-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:524]: Could you just for readability and completeness check that the collation of column c1 changed? Knowing that we passed the change, but not checking what we actually did is not sufficient.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 3]
```diff
@@ -516,6 +516,16 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
     }
   }
 
+  test("SPARK-48413: Alter column with collation") {
+    val tableName = "testcat.alter_column_tbl"
+    withTable(tableName) {
+      sql(s"CREATE TABLE $tableName (c1 STRING) USING PARQUET")
+      sql(s"ALTER TABLE $tableName ALTER COLUMN c1 TYPE STRING COLLATE UTF8_LCASE")
+      sql(s"INSERT INTO $tableName VALUES ('a')")
```
Comment (mihailom-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:524]:
```suggestion
      sql(s"INSERT INTO $tableName VALUES ('a')")
      sql(s"ALTER TABLE $tableName ALTER COLUMN c1 TYPE STRING COLLATE UTF8_LCASE")
```

Comment (mihailom-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:524]: Let's first insert and then alter, seems a bit nicer for the test.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 4]
```diff
@@ -516,6 +516,16 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
     }
   }
 
+  test("SPARK-48413: Alter column with collation") {
```
Comment (stefankandic) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:519]: can we also add tests for complex types (struct, map, array)?

**Review by stefankandic — COMMENTED** [review]
Please explain you change a bit more in the PR description

**Review by stefankandic — APPROVED** [review]

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @jovanm-db and @mihailom-db @stefankandic for review.

**Pull request closed without merging**